### ISSUE TITLE
Unexpected response on aliyun cloud provider metadata makes agent crash

### ISSUE THREAD

**Issue by superseb** [rancher/rancher#10474]
**Rancher versions:**
rancher/server: v1.6.12
rancher/agent: v1.2.7

**Setup details: (single node rancher vs. HA rancher, internal DB vs. external DB)**
Single node rancher, internal db

**Environment Template: (Cattle/Kubernetes/Swarm/Mesos)**
Cattle

**Steps to Reproduce:**
1. Launch agent in an environment with transparent anti malware/virus device intercepting requests to "rogue" IP ranges

For example, `curl -v http://100.100.100.200` will respond with an HTML page saying it's blocked.

**Results:**
```
{"log":"time=\"2017-12-05T14:24:18Z\" level=info msg=\"Launching agent\" \n","stream":"stdout","time":"2017-12-05T14:24:18.036383805Z"}
{"log":"time=\"2017-12-05T14:24:18Z\" level=info msg=\"Detecting cloud provider\" \n","stream":"stdout","time":"2017-12-05T14:24:18.03642389Z"}
{"log":"time=\"2017-12-05T14:24:18Z\" level=info msg=\"Listening for events on http://rancher:8080/v1\" \n","stream":"stdout","time":"2017-12-05T14:24:18.036567104Z"}
{"log":"time=\"2017-12-05T14:24:18Z\" level=info msg=\"launching hostapi\" \n","stream":"stdout","time":"2017-12-05T14:24:18.036679752Z"}
{"log":"time=\"2017-12-05T14:24:18Z\" level=warning msg=\"stat /var/lib/rancher/state/info.json: no such file or directory\" \n","stream":"stdout","time":"2017-12-05T14:24:18.036787447Z"}
{"log":"time=\"2017-12-05T14:24:18Z\" level=warning msg=\"stat /var/lib/rancher/state/info.json: no such file or directory\" \n","stream":"stdout","time":"2017-12-05T14:24:18.037365448Z"}
{"log":"panic: runtime error: index out of range\n","stream":"stderr","time":"2017-12-05T14:24:18.060848165Z"}
{"log":"\n","stream":"stderr","time":"2017-12-05T14:24:18.060883442Z"}
{"log":"goroutine 11 [running]:\n","stream":"stderr","time":"2017-12-05T14:24:18.060888529Z"}
{"log":"github.com/rancher/agent/vendor/github.com/denverdino/aliyungo/metadata.(*MetaData).Zone(0xc4202917e0, 0xc4200161c0, 0xc4200ae480, 0x0, 0x0)\n","stream":"stderr","time":"2017-12-05T14:24:18.060892832Z"}
{"log":"\u0009/go/src/github.com/rancher/agent/vendor/github.com/denverdino/aliyungo/metadata/client.go:229 +0xb0\n","stream":"stderr","time":"2017-12-05T14:24:18.060897477Z"}
{"log":"github.com/rancher/agent/cloudprovider/aliyun.metadataClientImpl.Zone(0xc4202917e0, 0x0, 0xc420033e60, 0x4ecbbb, 0xc4200161c0)\n","stream":"stderr","time":"2017-12-05T14:24:18.060902209Z"}
{"log":"\u0009/go/src/github.com/rancher/agent/cloudprovider/aliyun/aliyun_provider.go:43 +0x2b\n","stream":"stderr","time":"2017-12-05T14:24:18.060906763Z"}
{"log":"github.com/rancher/agent/cloudprovider/aliyun.(*Provider).GetHostInfo(0xc42024bb40, 0x0, 0x0, 0x0)\n","stream":"stderr","time":"2017-12-05T14:24:18.060911236Z"}
{"log":"\u0009/go/src/github.com/rancher/agent/cloudprovider/aliyun/aliyun_provider.go:57 +0x48\n","stream":"stderr","time":"2017-12-05T14:24:18.060915644Z"}
{"log":"github.com/rancher/agent/cloudprovider.GetCloudProviderInfo.func1(0xf4af80, 0xc42024bb40)\n","stream":"stderr","time":"2017-12-05T14:24:18.060920056Z"}
{"log":"\u0009/go/src/github.com/rancher/agent/cloudprovider/cloud_provider.go:60 +0x64\n","stream":"stderr","time":"2017-12-05T14:24:18.060924422Z"}
{"log":"created by github.com/rancher/agent/cloudprovider.GetCloudProviderInfo\n","stream":"stderr","time":"2017-12-05T14:24:18.060928857Z"}
{"log":"\u0009/go/src/github.com/rancher/agent/cloudprovider/cloud_provider.go:80 +0x1f5\n","stream":"stderr","time":"2017-12-05T14:24:18.060933144Z"}
```

Workaround is using `-e CATTLE_DETECT_CLOUD_PROVIDER=false` to the command running `rancher/agent`

**Comment by superseb** [comment]
@loganhz Can you verify this is fixed with the fix mentioned in https://github.com/rancher/rancher/issues/10439 ?

**Comment by loganhz** [comment]
@superseb I need to submit a PR to rancher/agent. This issue will be fixed after the rancher/agent change.

**Comment by superseb** [comment]
Fix for https://github.com/rancher/rancher/issues/10439 is for a non-200 return code. This issue is when the return code is 200 but the response is not valid for aliyun metadata.

**Comment by loganhz** [comment]
This is an Aliyun server side issue. I have added a recover logic to avoid rancher agent crash.

**Comment by deniseschannon** [comment]
@kingsd041 this can be tested with `rancher/server:v1.6-development` which pulls from master or you can wait for a RC (which should be monday)

**Comment by deniseschannon** [comment]
We were unable to reproduce this issue as it requires a aliyun server side crash. We've made a fix for it in v1.6.13. if this happens again after the release, please re-open.

**Comment by superseb** [comment]
This can be tested using:

```
ifconfig eth0:0 100.100.100.200 up
docker run -d --net=host -p 80:80 nginx
sudo docker run --rm --privileged -v /var/run/docker.sock:/var/run/docker.sock -v /var/lib/rancher:/var/lib/rancher rancher/agent:v1.2.8-rc1 http://xxx
```

Just tested, and it works.
